src/autodiff/gradfunc.jl

```diff
@@ -48,6 +48,41 @@ end
     (~protectf(g).f)(args...; kwargs...)
 end
 
+"""
+    gradient(Val(iloss), f, args::Tuple; kwargs...)
+    gradient(f, args; iloss::Int, kwargs...)
+    
+Calculate the gradient of `f(args...; kwargs...)` for reversible function `f` with regard to
+input `args`. The integer value `iloss` specifies the position of `loss` in `args`.
+
+!!! note
+    `iloss=1` is specially optimized, so putting the loss as the first parameter can avoid potential overhead.
+
+# Examples
+
+```jldoctest; setup=:(using NiLang)
+X = rand(2, 2)
+grads = NiLang.AD.gradient(Val(1), i_norm2, (0.0, X))
+grads[2] ≈ 2 .* X
+
+#output
+true
+```
+
+Note that `gradient` calculation is disabled for container with integers:
```

Comment:
I don't know why this is disabled. Is it possible to return `nothing` for it; otherwise it can very easily be a root of potential bugs.

Reply:
NiLang assumes gradient type == value type. Integers can be differentiated manually by wrapping GVar. Also, I assume in most cases, people do not want to differentiate them.
Wondering what are the potential bug?

Reply:
It is not very easy to detect and understand that `NiLang.AD.gradient(Val(1), i_sum, (0, X))[2]` is zeros (maybe throw a warning for this?)

Another example comes from the [Zygote example](https://giggleliu.github.io/NiLang.jl/dev/examples/port_zygote/), and it's quite confusing from the error stack:

```julia
julia> norm2'([1,2,3,4])
4-element Vector{Int64}:
 2
 4
 6
 8

julia> Zygote.@adjoint function norm2(x::AbstractArray{T}) where T
           out = norm2(x)
           out, δy -> (grad((~r_norm2)(GVar(out, δy), GVar(x))[2]),)
       end

julia> norm2'([1,2,3,4])
ERROR: MethodError: no method matching (::Inv{typeof(r_norm2)})(::GVar{Int64, Int64}, ::Vector{Int64})
Closest candidates are:
  (::Inv{typeof(r_norm2)})(::T, ::AbstractArray{T, N} where N) where T at REPL[8]:2
Stacktrace:
 [1] (::var"#10#11"{Vector{Int64}, Int64})(δy::Int64)
   @ Main ./REPL[11]:3
 [2] (::var"#138#back#12"{var"#10#11"{Vector{Int64}, Int64}})(Δ::Int64)
   @ Main ~/.julia/packages/ZygoteRules/OjfTt/src/adjoint.jl:59
 [3] (::Zygote.var"#41#42"{var"#138#back#12"{var"#10#11"{Vector{Int64}, Int64}}})(Δ::Int64)
   @ Zygote ~/.julia/packages/Zygote/6HN9x/src/compiler/interface.jl:41
```

Reply:
Here is how you can circumvent this issue:
```julia
julia> Zygote.@adjoint function norm2(x::AbstractArray{T}) where T; out = norm2(x); out, δy -> (grad((~r_norm2)(GVar(out, δy), GVar(x, zero(x)))[2]),) end

julia> norm2'([1,2,3,4])
4-element Vector{Int64}:
 2
 4
 6
 8
```

Does this make sense?